### PR TITLE
[Bug fix] unary: compare a preallocated output_tensor's shape against the input, not against itself

### LINKED ISSUE #55800: [Bug] ttnn unary ops accept a preallocated output_tensor of the wrong shape: the shape guard compares the shape against itself
### Describe the bug

The `TT_FATAL` that is supposed to reject a preallocated `output_tensor` whose shape does not match the input compares the preallocated shape **against itself**, so it can never fire. Every `ttnn` unary op therefore accepts a mismatched `output_tensor` silently: an oversized one makes the kernel read tiles from past the end of the input buffer, an undersized one silently returns a truncated result.

[`unary_device_operation.cpp#L101-L107`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp#L101-L107):

```cpp
if (output_tensor.has_value()) {
    const auto computed_output_shape = compute_output_specs(args, tensor_args).logical_shape();
    const auto preallocated_output_shape = output_tensor->logical_shape();
    TT_FATAL(
        preallocated_output_shape == computed_output_shape,
        "Unary: Preallocated output shape must match computed shape. ...");
```

and twenty lines below, [`compute_output_specs`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp#L118-L122) begins:

```cpp
tt::tt_metal::TensorSpec UnaryDeviceOperation::compute_output_specs(
    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
    if (tensor_args.output_tensor.has_value()) {
        return tensor_args.output_tensor->tensor_spec();
    }
```

The guard only runs when `output_tensor.has_value()`, and in exactly that case `compute_output_specs` returns the preallocated tensor's own spec — so the assertion is `output_tensor->logical_shape() == output_tensor->logical_shape()`. The layout check on the next line is written correctly (it compares against `input_tensor.layout()`), which is why a `ROW_MAJOR` output *is* rejected; only the shape arm is dead.

What gets past it is not harmless. The work split is sized from the **output**, [`unary_program_factory.cpp#L206-L207`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_program_factory.cpp#L206-L207):

```cpp
const uint32_t out_num_tiles =
    rm_interleaved ? (k.total_rows + k.rows_per_tile - 1) / k.rows_per_tile : output.physical_volume() / tile_hw;
```

so a 4-tile output makes the reader pull 4 pages out of a 1-page input buffer.

### Repro

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

def t(shape, val):
    return ttnn.from_torch(torch.full(shape, val, dtype=torch.float32),
                           dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)

x = t([1, 1, 32, 32], 2.0)          # 1 page
out = t([1, 1, 64, 64], -99.0)      # 4 pages, prefilled with a sentinel

r = ttnn.exp(x, output_tensor=out)  # expected: TT_FATAL. actual: runs.
print(r.shape)                                                  # Shape([1, 1, 64, 64])
print(sorted(set(ttnn.to_torch(out).flatten().tolist()))[:4])   # [1.0, 7.40625]

# 7.40625 = bf16 exp(2), in 1024 of the 4096 elements -- the one tile that exists.
# 1.0 = exp(0), in the other 3072 -- read from past the end of the 1-page input buffer.
# No element still holds the -99 sentinel, so all four tiles were written.

ttnn.add(x, x, output_tensor=out)   # control: binary_ng raises TT_FATAL here
```

### Observed vs expected

ttsim Blackhole v1.10.6, `bfloat16`, `TILE_LAYOUT`, input filled with `2.0`, `output_tensor` prefilled with the sentinel `-99.0`. This is a validation bug, not a numerics one, so the interesting column is what the caller's buffer holds rather than a ULP count.

| # | call | got | want |
|---|---|---|---|
| A | `ttnn.exp(x[1,1,32,32], output_tensor=out[1,1,64,64])` | no error; returns `[1,1,64,64]` backed by `out`; 1024/4096 elements are `7.40625` (`exp(2)`), 3072 are `1.0` (`exp(0)`, i.e. zeros read past the end of the 1-page input buffer), 0 still hold the sentinel | `TT_FATAL` "Unary: Preallocated output shape must match computed shape" |
| B | `ttnn.exp(x[1,1,64,64], output_tensor=out[1,1,32,32])` | no error; returns `[1,1,32,32]`, all 1024 elements `7.40625` — three quarters of the requested result silently dropped | same `TT_FATAL` |
| C | `ttnn.exp(x[1,1,32,64], output_tensor=out[1,1,64,32])` | no error; the returned tensor carries the caller's shape, not the input's | same `TT_FATAL` |
| D | `ttnn.relu(x[1,1,32,32], output_tensor=out[1,1,32,96])` | no error; all 3072 elements read `2.0` — the two out-of-range tiles came back as `2.0` in this run where they came back as `0.0` in run A | same `TT_FATAL` |
| E | `ttnn.add(x[1,1,32,32], x, output_tensor=out[1,1,64,64])` (control) | `TT_FATAL` from `binary_ng_device_operation.cpp` | correct today |

Run D is the point about the memory safety: the contents of the out-of-range tiles depend on unrelated allocations, not on the input, which is the signature of an out-of-bounds read rather than a defined value.

### Affected range

Every `ttnn` unary op that accepts `output_tensor=` — the whole `UnaryDeviceOperation` family (`exp`, `relu`, `sqrt`, `neg`, `abs`, `tanh`, `gelu`, `typecast`, …), every dtype, every layout. It is host-side device-op C++, so no kernel or SFPU code is involved.

The same idiom — validate against `compute_output_specs()`, which early-returns the preallocated tensor's spec — appears in four more device ops and is dead in all of them:

| file | line | verified |
|---|---|---|
| `ttnn/cpp/ttnn/operations/eltwise/ternary/device/ternary_device_operation.cpp` | 379 (guard), 464 (early return) | yes — `ttnn.where(x[1,1,32,32], x, x, output_tensor=out[1,1,64,64])` runs and returns a `[1,1,64,64]` tensor whose three extra tiles are read out of range |
| `ttnn/cpp/ttnn/operations/experimental/dropout/device/dropout_device_operation.cpp` | 69 (guard), 88 (early return) | by inspection |
| `ttnn/cpp/ttnn/operations/experimental/unary_backward/gelu_backward/device/gelu_backward_device_operation.cpp` | 98 (guard), 111 (early return) | by inspection |
| `ttnn/cpp/ttnn/operations/experimental/ssm/repeat_and_interleave_eltwise_mul/device/repeat_and_interleave_eltwise_mul_device_operation.cpp` | 79 (guard), 96 (early return) | by inspection |

`binary_ng` is the one family that gets it right, and only because it does the comparison *inside* `compute_output_specs`, against the freshly computed broadcast shape, before the early return.

### Model impact

None known, and stated up front: no in-tree model passes a mismatched `output_tensor` on purpose, so nothing in the repo is silently producing wrong numbers today. The value of fixing it is that a caller's shape mistake becomes the error the code already meant to raise instead of an out-of-bounds read whose result depends on what else is resident in DRAM.

### Expected behavior

A preallocated `output_tensor` whose logical shape differs from the shape the op would have produced should be rejected with the `TT_FATAL` that is already written, the same way `binary_ng` rejects it today. Unary ops are elementwise, so the expected shape is the input's logical shape — which is exactly what `compute_output_specs` computes when no output tensor is preallocated.

### Which architectures

All of them. This is host-side device-op validation in `ttnn/cpp`, not a kernel, so Wormhole, Blackhole and Quasar are affected identically.

### How it was verified

ttsim Blackhole v1.10.6 (`TT_METAL_SIMULATOR=libttsim_bh.so`, slow dispatch), `bfloat16`, `TILE_LAYOUT`, with the `output_tensor` prefilled with `-99.0` so that "written" and "left alone" can be told apart, and `buffer_address()` compared to confirm the returned tensor is the caller's buffer. `ttnn.add` and a `ROW_MAJOR` output were used as controls to show the surrounding validation does fire.

### PR DESCRIPTION
**Pull request by badnikhil**
### PR Category

Bug fix

### Summary

Closes #55800

`UnaryDeviceOperation::validate_on_program_cache_miss` took the expected output shape from `compute_output_specs()`, and `compute_output_specs()` starts with

```cpp
if (tensor_args.output_tensor.has_value()) {
    return tensor_args.output_tensor->tensor_spec();
}
```

The guard only runs when `output_tensor.has_value()`, so in exactly the case it is meant to police the "computed" shape *is* the preallocated tensor's own shape. The `TT_FATAL` compared the tensor against itself and could never fire.

A mismatched `output_tensor` was therefore accepted silently by every `ttnn` unary op. The work split is sized from the output (`unary_program_factory.cpp:206`, `output.physical_volume() / tile_hw`), so an oversized one makes the reader pull pages from past the end of the input buffer, and an undersized one silently returns a truncated result.

Unary ops are elementwise, so the shape the op would have produced is the input's logical shape — exactly what `compute_output_specs` computes on the branch where nothing is preallocated. One line, plus a comment saying why `compute_output_specs()` must not be used here. No SFPU operations added; nothing on the fast path changes, and nothing changes at all for a correctly shaped `output_tensor`.

ttsim Blackhole v1.10.6, bfloat16, TILE layout, input filled with `2.0`, `output_tensor` prefilled with the sentinel `-99.0`:

| call | before | after |
|---|---|---|
| `ttnn.exp(x[1,1,32,32], output_tensor=out[1,1,64,64])` | no error; returns `[1,1,64,64]`; 1024/4096 elements `7.40625` (`exp(2)`), 3072 `1.0` (`exp(0)` — zeros read past the end of the 1-page input buffer), 0 sentinel left | **`TT_FATAL` "Unary: Preallocated output shape must match computed shape. Computed: Shape([1, 1, 32, 32]), Preallocated: Shape([1, 1, 64, 64])"** |
| `ttnn.exp(x[1,1,64,64], output_tensor=out[1,1,32,32])` | no error; returns `[1,1,32,32]`, all `7.40625` — three quarters of the result dropped | **`TT_FATAL`** |
| `ttnn.exp(x[1,1,32,64], output_tensor=out[1,1,64,32])` | no error; returned tensor carries the caller's shape, not the input's | **`TT_FATAL`** |
| `ttnn.relu(x[1,1,32,32], output_tensor=out[1,1,32,96])` | no error; all 3072 elements `2.0` — the two out-of-range tiles read `2.0` here and `0.0` in the run above, i.e. whatever was resident | **`TT_FATAL`** |
| `ttnn.add(x[1,1,32,32], x, output_tensor=out[1,1,64,64])` (control) | `TT_FATAL` from `binary_ng` | unchanged |
| `ttnn.exp(x[1,1,32,64], output_tensor=out[1,1,32,64])` (matching) | runs, writes `out` | unchanged |

unary preallocated output shape guard, before/after

Raw text: https://raw.githubusercontent.com/badnikhil/tt-metal/pr-assets/tt-metal-unary-outshape.txt

`tests/ttnn/unit_tests/operations/eltwise/test_unary.py`, ttsim Blackhole, whole file:

| | main | with this change |
|---|---|---|
| `-k preallocated_output` (the new tests) | 17 failed, 1 passed | **18 passed** |
| whole file | 838 passed, 17 failed, 3 xfailed | **855 passed, 3 xfailed** |

The 17 failures on `main` are the new tests' mismatch cases; there are no other failures in either run. Four cases are deselected in both columns (`h=262144`, 8.4M elements — they exhaust the 13 GB machine this was run on, not a code issue).

### Notes for reviewers

- The layout check three lines below was already written correctly — it compares against `input_tensor.layout()` — which is why a `ROW_MAJOR` `output_tensor` *is* rejected today and only the shape arm was dead. It is untouched.
- The comparison stays a plain `Shape ==`, matching the guard's existing wording and matching `ternary` and `dropout`. `binary_ng` is looser (it ignores leading-1 rank differences) because broadcasting inherently changes rank; unary has no such case.
- `test_unary_preallocated_output_shape_mismatch` covers 5 mismatches (oversized, undersized, same volume with a different shape, a non-multiple page count, and a dropped batch dimension) across `exp`, `relu` and `sqrt`. All 15 pass silently on `main` and raise with this change. `test_unary_preallocated_output_shape_match` is the other direction — a correctly shaped preallocated output must still run and still write the caller's buffer — and passes both before and after.
- No behaviour change for any correctly shaped `output_tensor`, and none at all when `output_tensor` is not supplied: the branch is inside `if (output_tensor.has_value())`.
- The check lives in `compute_output_specs`, the way `binary_ng` does it, and `validate_on_program_cache_miss` calls `compute_output_specs` for the preallocated case so it still runs with the program cache disabled. Unary defines no `validate_on_program_cache_hit`, so the framework routes cache hits to the cache-miss validator and the check fires there too; `test_unary_preallocated_output_shape_mismatch_program_cache_hit` and `..._program_cache_disabled` cover those two paths.
- No in-tree model passes a mismatched `output_tensor` on purpose, so nothing in the repo is silently producing wrong numbers today; what this buys is that a caller's shape mistake becomes the error the code already meant to raise instead of an out-of-bounds read.
- The same idiom (validate against `compute_output_specs()`, which early-returns the preallocated spec) is dead in four more device ops — `ternary_device_operation.cpp:379`, `dropout_device_operation.cpp:69`, `gelu_backward_device_operation.cpp:98`, `repeat_and_interleave_eltwise_mul_device_operation.cpp:79`. `ttnn.where` with a mismatched `output_tensor` is confirmed to run on ttsim, same as unary. They are listed in the issue and left out of this PR to keep it to one op family and one test file; happy to fold them in here if you would rather have them together.
- Verified on ttsim Blackhole v1.10.6 (slow dispatch); no silicon. This is host-side device-op validation, so it is arch-independent.

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py [hunk 1]
```diff
@@ -2758,3 +2758,48 @@ def test_softcap_zero_beta_guard(device, expect_error):
...
+    with expect_error(RuntimeError, "Preallocated output shape must match computed shape"):
+        ttnn_function(input_tensor, output_tensor=output_tensor)
+
+
+def test_unary_preallocated_output_shape_match(device):
+    """The matching case still runs and still writes the caller's buffer."""
+    shape = [1, 1, 32, 64]
+    torch_input = torch.linspace(-2.0, 2.0, 32 * 64, dtype=torch.bfloat16).reshape(shape)
+
+    input_tensor = ttnn.from_torch(torch_input, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
```
Comment (mateusznowakTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary.py:2818]: Are there cases other than ones reaching validate_on_program_cache_miss that should assert if output_tensor is incorrectly provided? If so, those shall be also tested.

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp [hunk 2]
```diff
@@ -99,7 +99,12 @@ void UnaryDeviceOperation::validate_on_program_cache_miss(
     }
 
     if (output_tensor.has_value()) {
-        const auto computed_output_shape = compute_output_specs(args, tensor_args).logical_shape();
```
Comment (mateusznowakTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:102]: Please follow the binary_ng pattern where such checks are done inside compute_output_specs.

**Review by mateusznowakTT — CHANGES_REQUESTED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp (hunk 2 above)
Comment (badnikhil) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:102]: Moved into compute_output_specs, before the preallocated spec is returned. validate_on_program_cache_miss now just calls compute_output_specs for the preallocated case, so the check still runs when the program cache is disabled and compute_program_hash is never reached.

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py (hunk 1 above)
Comment (badnikhil) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary.py:2818]: Unary defines no validate_on_program_cache_hit, so the framework routes cache hits back to validate_on_program_cache_miss, and graph capture goes through the same call. Every dispatch path reaches the check. Added the two that are not obvious as tests: a program cache hit (compute_program_hash does not hash the output shape, so a mismatched call after a matching one is a hit) and a run with the program cache disabled. Both fail on main and pass with the fix. Whole file on ttsim Blackhole: 855 passed, 3 xfailed.

ttnn.where still accepts a mismatched output_tensor for the same reason. It is in the issue table and deliberately not in this PR.

**Review by mateusznowakTT — APPROVED** [review]

**Comment by mateusznowakTT** [comment]
Sanity https://github.com/tenstorrent/tt-metal/actions/runs/35203284753 and Nightly L2 https://github.com/tenstorrent/tt-metal/actions/runs/35203314442 triggered for the change - failures are unrelated.
